pkg/catalog/catalog.go

```diff
@@ -1838,12 +1832,62 @@ func (c *Catalog) writeUncommittedLocal(ctx context.Context, repository *gravele
 	if itr.Err() != nil {
 		return nil, false, itr.Err()
 	}
-	err = pw.WriteStop()
+
+	// Finished reading all staging area - no continuation token
+	hasData := count > 0
+	return nil, hasData, nil
+}
+
+func (c *Catalog) writeTokensLocal(ctx context.Context, repository *graveler.RepositoryRecord, w *UncommittedWriter, pw *writer.ParquetWriter, mark *GCUncommittedMark, runID string) (*GCUncommittedMark, bool, error) {
+	count := 0
+	itr, err := c.Store.ListAddressTokens(ctx, repository)
 	if err != nil {
 		return nil, false, err
 	}
+	defer itr.Close()
 
-	// Finished reading all staging area - no continuation token
+	if mark != nil && mark.BranchID == "" {
+		itr.SeekGE(mark.Path.String())
+	}
+
+	for itr.Next() {
```

Comment:
Consider reusing the iteration logic and just providing an iterator and a write function to it